src/routes/authorize.rs

```diff
@@ -33,49 +39,99 @@ async fn authorize_handler(
+            query.state.as_deref(),
+        );
+    };
+
+    // `response_type` is required.
+    let response_type: &str = if let Some(response_type) = &query.response_type {
+        response_type
+    } else {
+        return error_redirect(redirect_uri, Error::ResponseType, query.state.as_deref());
+    };
+
+    let client_configs = if let Some(configs) = app_state.client_configs.get(&client_id) {
+        configs
+    } else {
+        return error_redirect(
+            Url::parse(&query.redirect_uri).unwrap(),
+            Error::OauthInvalidClientId,
+            query.state.as_deref(),
+        );
+    };
+
+    let redirect_urls: &Vec<url::Url> = &client_configs.redirect_urls;
+    let requirements_empty = client_configs.requirements.is_empty();
 
     // Support Authorization Code Flow and Implicit Flow.
-    let response_type = ResponseType::from_str(query.response_type.as_str())?;
+    let response_type = ResponseType::from_str(response_type)?;
 
     // Implicit flow must include a nonce.
     if response_type.is_implicit_flow() && query.nonce.is_none() {
-        return Err(Error::InvalidNonce);
+        return error_redirect(redirect_uri, Error::InvalidNonce, query.state.as_deref());
     }
 
-    let is_redirect_uri_in_query = redirect_urls
-        .contains(&Url::parse(&query.redirect_uri).map_err(|_| Error::OauthInvalidRedirectUri)?);
+    let is_redirect_uri_in_query = redirect_urls.contains(&redirect_uri);
     if !is_redirect_uri_in_query {
-        return Err(Error::OauthInvalidRedirectUri);
+        return error_redirect(
+            redirect_uri,
+            Error::OauthInvalidRedirectUri,
+            query.state.as_deref(),
+        );
     }
 
-    session.insert(REDIRECT_URI_SESSION_KEY, query.redirect_uri.clone())?;
-    session.insert(RESPONSE_TYPE_SESSION_KEY, query.response_type.clone())?;
+    let authorize_parameters = AuthorizeParameters {
+        client_id,
+        redirect_uri,
+        response_type,
+        scope: query.scope.clone().unwrap_or_default(),
+        state: query.state.clone(),
+        nonce: query.nonce.clone(),
+    };
 
     match (requirements_empty, &query.nonce) {
         (true, Some(nonce)) => {
-            session.insert(OIDC_SESSION_KEY, query.clone().into_inner())?;
+            session.insert(OIDC_SESSION_KEY, authorize_parameters)?;
             let redirect_uri_with_nonce = format!("/?nonce={}", nonce);
             Ok(HttpResponse::Found()
                 .append_header(("Location", redirect_uri_with_nonce))
                 .finish())
         }
         _ => {
-            session.insert(OIDC_SESSION_KEY, query.clone().into_inner())?;
+            session.insert(OIDC_SESSION_KEY, authorize_parameters)?;
             Ok(HttpResponse::Found()
                 .append_header(("Location", "/"))
                 .finish())
         }
     }
 }
+
+/// Creates a 302 response with an error.
+fn error_redirect(
+    mut redirect_uri: Url,
```

Comment:
I don't think I've ever seen a `mut` taken directly. I think it whould be either `&mut` or then if you take ownership of the object, it does not need to be `mut`, at least not in its signature.

Reply:
Well, we want to take ownership of the variable, hence no reference. And we want to mutate it hence the `mut`. if you don't add it the compiler will not let you mutate it.